File: cppDir/HandAlyzer.cpp
#include <iostream>
#include <algorithm>
#include <sstream>
#include <iomanip>
#include "HandAlyzer.h"



using namespace std;


// implement the HandAlyzer class

//---------------------------------------------------------//
//Data implementation

    
    //For tracking win and tie breaking info
    //The following are what each index is used for
    //[0] = HandRank //denoted as 0 (high card)) -> 9 (R straight flush)
    //[1] = Highest Rank card (or highest Rank of pair)
    //[2] = 2nd highest Rank card (if needed, else 0)
    //[3] = 3rd highest Rank card (if needed, else 0)
    //[4] = 4th highest Rank card (if needed, else 0)
    //[5] = 5th highest Rank card (if needed, else 0)
    //[6] = Suit of card in [1] // denoted as (0, D) (1, C) (2, H) (3, S)

    int scoreKeeperAll[6][7] = {{0, 0, 0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0, 0},
                                {0, 0, 0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0, 0}
                            };

    // Constructor
    HandAlyzer::HandAlyzer(Card handsImport[][5]) {
        //Can only be constructed with [6 player hands][5 Cards a hand]

        //update players1-6 hands#[]
        for (int i = 0; i < 5; i++) {
            hand1[i] = handsImport[0][i];
            hand2[i] = handsImport[1][i];
            hand3[i] = handsImport[2][i];
            hand4[i] = handsImport[3][i];
            hand5[i] = handsImport[4][i];
            hand6[i] = handsImport[5][i];


        }
        //update master hands[][]
        for (int i = 0; i < 6; i++) {
            for (int j = 0; j < 5; j++){
                hands[i][j] = handsImport[i][j];
            }
        }
    }
//---------------------------------------------------------//
//Methods

    //Converts HandRank enum to string
    string HandAlyzer::handRanktoString(HandRank rank) { 
        string rankString = "";
        switch (rank) {
            case HIGH_CARD: rankString += "High Card"; break;
            case PAIR: rankString += "Pair"; break;
            case TWO_PAIR: rankString += "Two Pair"; break;
            case THREE_OF_A_KIND: rankString += "Three of a Kind"; break;
            case STRAIGHT: rankString += "Straight"; break;
            case FLUSH: rankString += "Flush"; break;
            case FULL_HOUSE: rankString += "Full House"; break;
            case FOUR_OF_A_KIND: rankString += "Four of a Kind"; break;
            case STRAIGHT_FLUSH: rankString += "Straight Flush"; break;
            case ROYAL_STRAIGHT_FLUSH: rankString += "Royal Straight Flush"; break;
        }
        return rankString;
    }

    //Converts HandRank enum to int
    int HandAlyzer::handRanktoInt(HandRank rank) { 
        int rankInt = 0;
        switch (rank) {
            case HIGH_CARD: rankInt = 0; break;
            case PAIR: rankInt = 1; break;
            case TWO_PAIR: rankInt = 2; break;
            case THREE_OF_A_KIND: rankInt = 3; break;
            case STRAIGHT: rankInt = 4; break;
            case FLUSH: rankInt = 5; break;
            case FULL_HOUSE: rankInt = 6; break;
            case FOUR_OF_A_KIND: rankInt = 7; break;
            case STRAIGHT_FLUSH: rankInt = 8; break;
            case ROYAL_STRAIGHT_FLUSH: rankInt = 9; break;
        }
        return rankInt;
    }



    //------------------------------------------------//
    void HandAlyzer::getFinalScorePrint() {
        string handStrings[6];

        //Get hand rank and build strings for printing
        for (int i = 0; i < 6; i++) {
            int scoreKeeper[7] = {0};
            HandRank rank = getHandRank(hands[i], scoreKeeper);
            playerHandRanks[i] = rank;
            scoreKeeper[0] = handRanktoInt(rank);

            //updates master scoreKeeper
            for (int j = 0; j < 7; j++){
                scoreKeeperAll[i][j] = scoreKeeper[j]; 
            }

            //Build out Hands to print, stored in order of players 1->6
            stringstream handStringStream;
            for (int j = 0; j < 5; j++) {
                handStringStream << setw(4) << hands[i][j].toString();
            }
            
            // //DEBUGGING: Adds on to print statement to show scoreKeeper values
            // for (int j = 0; j < 7; j++) {
            //     if (scoreKeeper[j] < 10){
            //         handStringStream << " [0" << scoreKeeper[j] << "] ";
            //     } 
            //     else {handStringStream << " [" << scoreKeeper[j] << "] ";}
            // }
            // //DEBUGGING: End.

            handStrings[i] = handStringStream.str();
        }

        ////Sort out winners and implement tie breaking via bubblesorting
        for (int playerRankIndexLimit = 0; playerRankIndexLimit < 6 ;  playerRankIndexLimit++){
            for(int currentPlayer = 0; currentPlayer < 5; currentPlayer++){
                if (scoreKeeperAll[currentPlayer][0] < scoreKeeperAll[currentPlayer + 1][0]){
                    swapHands(currentPlayer, currentPlayer + 1, handStrings);
                } 
                //Tie Breaking
                else if (scoreKeeperAll[currentPlayer][0] == scoreKeeperAll[currentPlayer + 1][0]){
                    for (int criteriaIndex = 1; criteriaIndex < 7; criteriaIndex++){
                        if(scoreKeeperAll[currentPlayer][criteriaIndex] < scoreKeeperAll[currentPlayer + 1][criteriaIndex]){
                            swapHands(currentPlayer, currentPlayer + 1, handStrings);
                            break;

                        } else if (scoreKeeperAll[currentPlayer][criteriaIndex] > scoreKeeperAll[currentPlayer + 1][criteriaIndex]){
                            break;
                        }
                    }
                }
            }
        }

        //Print the sorted hands with their ranks
        cout << "--- WINNING HAND ORDER ---" << endl;
        for (int i = 0; i < 6; i++) {
            cout << handStrings[i] << " - " << handRanktoString(playerHandRanks[i]) << endl;
        }

    }
    //-------------------
    // Helper method to swap hands and corresponding data
    void HandAlyzer::swapHands(int player1, int player2, string handStrings[]) {
        
        //Swaps scoreKeeper[p1] and [p2]
        for (int i = 0; i < 7; i++){
            int tempScoreKeeper = scoreKeeperAll[player1][i];
            scoreKeeperAll[player1][i] = scoreKeeperAll[player2][i];
            scoreKeeperAll[player2][i] = tempScoreKeeper;
        }

        HandRank tempRank = playerHandRanks[player1];
        playerHandRanks[player1] = playerHandRanks[player2];
        playerHandRanks[player2] = tempRank;
        
        string tempHand = handStrings[player1];
        handStrings[player1] = handStrings[player2];
        handStrings[player2] = tempHand;
    }
    //------------------------------------------------//
    HandRank HandAlyzer::getHandRank(Card hand[], int scoreKeeper[]) {
        //Checks one hand of 5 cards for its HandRank
        //Currently, assigns High Card if no other HandRank fits, then
        //uses getHighCardTieBreakers to determine ties
        //...
        //Other methods have independent means to adjust scoreKeeper
        //to suit the needs for tie breaking of equal strength hands

        if (isStraightRoyalFlush(hand, scoreKeeper)) {
            scoreKeeper[0] = 9;
            return ROYAL_STRAIGHT_FLUSH;
        } else if (isStraightFlush(hand, scoreKeeper)) {
            scoreKeeper[0] = 8;
            return STRAIGHT_FLUSH;
        } else if (isFourOfAKind(hand, scoreKeeper)) {
            scoreKeeper[0] = 7;
            return FOUR_OF_A_KIND;
        } else if (isFullHouse(hand, scoreKeeper)) {
            scoreKeeper[0] = 6;
            return FULL_HOUSE;
        } else if (isFlush(hand, scoreKeeper)) {
            scoreKeeper[0] = 5;
            return FLUSH;
        } else if (isStraight(hand, scoreKeeper)) {
            scoreKeeper[0] = 4;
            return STRAIGHT;
        } else if (isThreeOfAKind(hand, scoreKeeper)) {
            scoreKeeper[0] = 3;
            return THREE_OF_A_KIND;
        } else if (numOfPairs(hand, scoreKeeper) == 2) {
            scoreKeeper[0] = 2;
            return TWO_PAIR;
        } else if (numOfPairs(hand, scoreKeeper) == 1) {
            scoreKeeper[0] = 1;
            return PAIR;
        } else if (scoreKeeper[0] == 0) {
            getHighCard(hand, scoreKeeper);
            getHighCardTieBreakers(hand, scoreKeeper);
        }       
        return HIGH_CARD;
    }

    //------------------------------------------------//
    //Hand Rank Identifiers---------------------------//
    //------------------------------------------------//
    bool HandAlyzer::isStraight(Card hand[], int scoreKeeper[]) {
        int ranks[5] = {0};
        Card highestCard;

        for (int i = 0; i < 5; i++) {
            ranks[i] = hand[i].getRankInt();

            if (hand[i].getRankInt() > highestCard.getRankInt()) {
                highestCard = hand[i];
            }
        }

        sort(ranks, ranks + 5);

        if (isStraightRoyal(hand, scoreKeeper)) {
            return true;
        } else {
            for (int i = 0; i < 4; i++) {
                if (ranks[i] + 1 != ranks[i + 1]) {
                    return false;
                }
            }
        }

        getHighCard(hand, scoreKeeper);
        return true;
    }

    //------------------------------------------------//
    bool HandAlyzer::isFlush(Card hand[], int scoreKeeper[]) {
        for (int i = 0; i < 4; i++) {
            if (hand[i].getSuitInt() != hand[i + 1].getSuitInt()) {
                return false;
            }
        }
        getHighCard(hand, scoreKeeper);
        getHighCardTieBreakers(hand, scoreKeeper);
        return true;
    }

    //------------------------------------------------//
    //This method does NOT account for wrap around flushes!
    bool HandAlyzer::isStraightFlush(Card hand[], int scoreKeeper[]) {
        return (isFlush(hand, scoreKeeper) && isStraight(hand, scoreKeeper));                                                                                                                                                                                
    }

    //------------------------------------------------//
    bool HandAlyzer::isStraightRoyalFlush(Card hand[], int scoreKeeper[]) {
        int loopCounter = 0;
        int countsPerRank[13] = {0};
        int suitPerCard[5] = {0};
        //Array with hard coded rank values of A, 10, J, Q, K
        int cardsToCheck[5] = {0, 9, 10,
                              11, 12};

         for (int i = 0; i < 5; i++) {
            Card card = hand[i];
            countsPerRank[card.getRankInt()] += 1;
            suitPerCard[loopCounter] = card.getSuitInt();
            loopCounter++;
        }
        for (int i : cardsToCheck) {
            if (countsPerRank[i] != 1) {
                return false;
            }
        }

        for (int i = 0; i < 4; i++) {
            if (suitPerCard[i] != suitPerCard[i + 1]) {
                return false;
            }
        }
        
        scoreKeeper[6] = hand[0].getSuitInt();        
        return true;
    }

    //------------------------------------------------//
    //To solve edge case where (non-flush, royal) Straights were not identified
    //Catches straights in order of "10, J, Q, K, A" not caught by isStraight
    bool HandAlyzer::isStraightRoyal(Card hand[], int scoreKeeper[]) {
        int countsPerRank[13] = {0};
        //Array with hard coded int values for A, 10, J, Q, K
        int cardsToCheck[5] = {0, 9, 10,
                              11, 12};
        int aceSuit = 0;

         for (int i = 0; i < 5; i++) {
            Card card = hand[i];
            countsPerRank[card.getRankInt()] += 1;
            if (card.getRankInt() == 0) aceSuit = card.getSuitInt();                
        }

        for (int i : cardsToCheck) {
            if (countsPerRank[i] != 1) {
                return false;
            }
        }

        scoreKeeper[1] = 14; //Ace High value hard coded, since it will be in every SRF
        scoreKeeper[6] = aceSuit;       
        return true;
    }
    //------------------------------------------------//
    //Returns int for num of pairs, unlike rest returning bool!
    int HandAlyzer::numOfPairs(Card hand[], int scoreKeeper[]) {
        int numofPairs = 0;
        int iterator = 0;
        int kickerRank = 0;
        int rankofPair[2];
        int countsPerRank[13] = {0};

        for (int i = 0; i < 5; i++) {
            countsPerRank[hand[i].getRankInt()] += 1;
        }

        for (int i = 0; i < 13; i++) {
            if (countsPerRank[i] == 2) {
                numofPairs++;
                //check for aces, make high
                if (i == 0){
                    rankofPair[iterator] = 14;
                } else rankofPair[iterator] = i;

                iterator++;
            }
            //updates kicker if no pair, always ends with highest non-pair's rank
            else if (countsPerRank[i] == 1) {
                kickerRank = i;
            }
        }

       

        if (numofPairs == 1) {
            //update scorekeeping if only 1 pair
            int ranks[5] = {0};
            for (int i = 0; i < 5; i++) {
                ranks[i] = hand[i].getRankInt();
                //For ties, ranks ace as highest value
                if (ranks[i] == 0){
                    ranks[i] = 14;
                }
            }

            sort(ranks, ranks + 5); //sorted [low to high]
            int sKIndex = 2; //scoreKeeper[sKIndex], [2]->[5] is the 2nd highest card -> lowest card

            //Reverses the array from [low to high], to, [high to low] when assigning values to scoreKeeper[]
            //Only assigns value if the rank does not match the pair
            for (int i = 4; i >= 0; i--) {
                if (ranks[i] != rankofPair[0]){
                    scoreKeeper[sKIndex] = ranks[i];
                    sKIndex++;
                }
            }
            
            //FINISHED UPDATE
            scoreKeeper[1] = rankofPair[0];

        } else if (numofPairs == 2) {
            scoreKeeper[1] = rankofPair[1]; //higher value pair first
            scoreKeeper[2] = rankofPair[0]; //lower value pair second
        }
        
        if (numofPairs > 0){
          //finds kicker card, then updates scorekeeper[6] with suit
          for (int i = 0; i < 5; i++) {
              Card card = hand[i];
              if (card.getRankInt() == kickerRank) {
                  scoreKeeper[6] = card.getSuitInt();
              }
          }
          
        }

        return numofPairs;
    }

    //------------------------------------------------//
    bool HandAlyzer::isThreeOfAKind(Card hand[], int scoreKeeper[]) {
        int countsPerRank[13] = {0};
        for (int i = 0; i < 5; i++) {
            Card card = hand[i];
            countsPerRank[card.getRankInt()] += 1;
        }

        for (int i = 0; i < 13; i++) {
            if (countsPerRank[i] == 3) {
                getHighCardTieBreakers(hand, scoreKeeper);
                getHighCard(hand, scoreKeeper);
                return true;
            }
        }
        return false;
    }

    //------------------------------------------------//
    bool HandAlyzer::isFullHouse(Card hand[], int scoreKeeper[]) {
        if (isThreeOfAKind(hand, scoreKeeper) && (numOfPairs(hand, scoreKeeper) == 1)) {
            getHighCard(hand, scoreKeeper);
            return true; 
        }
        return false;
    }

    //------------------------------------------------//
    bool HandAlyzer::isFourOfAKind(Card hand[], int scoreKeeper[]) {
        int countsPerRank[13] = {0};
        for (int i = 0; i < 5; i++) {
            Card card = hand[i];
            countsPerRank[card.getRankInt()]++;
        }

        for (int i = 0; i < 13; i++) {
            if (countsPerRank[i] == 4) {
                getHighCard(hand, scoreKeeper);
                return true;
            }
        }
        return false;
    }

    Card HandAlyzer::getHighCard(Card playerHand[5], int scoreKeeper[7]){
        Card highCard;
            bool isAce = false;
            for (int i = 0; i < 5; i++) {
                Card cardToBeChecked = playerHand[i];

                //check for aces, if found only looks for other possible aces to compare
                if (cardToBeChecked.getRankInt() == 0){
                    if (highCard.getRankInt() != 0){
                        highCard = cardToBeChecked;
                    } 
                    else if (cardToBeChecked.getSuitInt() > highCard.getSuitInt()){
                        highCard = cardToBeChecked;
                    }
                    isAce = true;
                }
                //else, find highest rank card
                else if (highCard.getRankInt() < cardToBeChecked.getRankInt())
                {
                    highCard = cardToBeChecked;
                }
                //if tied for rank, compare suit
                else if ((highCard.getRankInt() == cardToBeChecked.getRankInt()) 
                        && (highCard.getSuitInt() < cardToBeChecked.getSuitInt()))
                {
                    highCard = cardToBeChecked;
                }
    
            }   
    
            //Keeps Aces High for high card
            if (isAce){
                scoreKeeper[1] = 14;
            } else scoreKeeper[1] = highCard.getRankInt();
    
            scoreKeeper[6] = highCard.getSuitInt();
    
            return highCard;
    
    }


    void HandAlyzer::getHighCardTieBreakers(Card hand[5], int scoreKeeper[7]){
        int ranks[5];
        for (int i = 0; i < 5; i++) {
            ranks[i] = hand[i].getRankInt();
            
           // //DEBUGGING: check to see if hand[] is being passed in right
           // cout << " DEBUG DEBUG :: " << hand[i].getRankInt() << endl;
           // 
           
            //For tiebreaking later, ranks ace as highest value
            if (ranks[i] == 0){
                ranks[i] = 14;
            }
        }
        sort(ranks, ranks + 5); //sorted [low to high]
        int sKIndex = 2;

        //Assigns the scoreKeeper[][][3][2][1][0][]
        //Where 3-0 is:
        //ranks[3] ~ 2nd highest card, to , ranks[0] ~ lowest card
        for (int i = 3; i >= 0; i--) {
            scoreKeeper[sKIndex] = ranks[i];
            sKIndex++;
        }


    }
    //------------------------------------------------//

File: julDir/HandAlyzer.jl
include("Deck.jl")
module HandAlyzerModule

export HandRank, HandAlyzer




# Enum for HandRank
@enum HandRank HIGH_CARD=0 PAIR TWO_PAIR THREE_OF_A_KIND STRAIGHT FLUSH FULL_HOUSE FOUR_OF_A_KIND STRAIGHT_FLUSH ROYAL_STRAIGHT_FLUSH

mutable struct HandAlyzer
    hands::Vector{Vector{Card}}
    playerHandRanks::Vector{HandRank}

    # Constructor
    function HandAlyzer(handsImport::Matrix{Card})
        new(Vector{Card}(undef, 6, 5), HandRank[])
        
        # update players1-6 hands#[]
        for i in 1:5
            hands[1][i] = handsImport[1, i]
            hands[2][i] = handsImport[2, i]
            hands[3][i] = handsImport[3, i]
            hands[4][i] = handsImport[4, i]
            hands[5][i] = handsImport[5, i]
            hands[6][i] = handsImport[6, i]
        end

        # update master hands[][]
        for i in 1:6
            for j in 1:5
                hands[i][j] = handsImport[i, j]
            end
        end
    end

    # Methods
    function handRanktoString(rank::HandRank)::String
        return switch(rank) do r
            HIGH_CARD => "High Card"
            PAIR => "Pair"
            TWO_PAIR => "Two Pair"
            THREE_OF_A_KIND => "Three of a Kind"
            STRAIGHT => "Straight"
            FLUSH => "Flush"
            FULL_HOUSE => "Full House"
            FOUR_OF_A_KIND => "Four of a Kind"
            STRAIGHT_FLUSH => "Straight Flush"
            ROYAL_STRAIGHT_FLUSH => "Royal Straight Flush"
        end
    end

    function handRanktoInt(rank::HandRank)::Int
        return Int(rank)
    end

    function getFinalScorePrint()
        handStrings = Vector{String}(undef, 6)

        # Get hand rank and build strings for printing
        for i in 1:6
            scoreKeeper = zeros(Int, 7)
            rank = getHandRank(hands[i], scoreKeeper)
            playerHandRanks[i] = rank
            scoreKeeper[1] = handRanktoInt(rank)

            # updates master scoreKeeper
            scoreKeeperAll[i, :] .= scoreKeeper

            # Build out Hands to print, stored in order of players 1->6
            handStringStream = IOBuffer()
            for j in 1:5
                print(handStringStream, lpad(string(hands[i][j]), 4))
            end
            handStrings[i] = String(take!(handStringStream))
        end

        # Sort out winners and implement tie breaking via bubblesorting
        for playerRankIndexLimit in 1:6
            for currentPlayer in 1:5
                if scoreKeeperAll[currentPlayer, 1] < scoreKeeperAll[currentPlayer + 1, 1]
                    swapHands(currentPlayer, currentPlayer + 1, handStrings)
                elseif scoreKeeperAll[currentPlayer, 1] == scoreKeeperAll[currentPlayer + 1, 1]
                    for criteriaIndex in 2:7
                        if scoreKeeperAll[currentPlayer, criteriaIndex] < scoreKeeperAll[currentPlayer + 1, criteriaIndex]
                            swapHands(currentPlayer, currentPlayer + 1, handStrings)
                            break
                        elseif scoreKeeperAll[currentPlayer, criteriaIndex] > scoreKeeperAll[currentPlayer + 1, criteriaIndex]
                            break
                        end
                    end
                end
            end
        end

        # Print the sorted hands with their ranks
        println("--- WINNING HAND ORDER ---")
        for i in 1:6
            println(handStrings[i], " - ", handRanktoString(playerHandRanks[i]))
        end
    end

    # Helper method to swap hands and corresponding data
    function swapHands(player1, player2, handStrings)
        scoreKeeperAll[player1, :], scoreKeeperAll[player2, :] = scoreKeeperAll[player2, :], scoreKeeperAll[player1, :]
        playerHandRanks[player1], playerHandRanks[player2] = playerHandRanks[player2], playerHandRanks[player1]
        handStrings[player1], handStrings[player2] = handStrings[player2], handStrings[player1]
    end

    function getHandRank(hand, scoreKeeper)
        if isStraightRoyalFlush(hand, scoreKeeper)
            scoreKeeper[1] = 9
            return ROYAL_STRAIGHT_FLUSH
        elseif isStraightFlush(hand, scoreKeeper)
            scoreKeeper[1] = 8
            return STRAIGHT_FLUSH
        elseif isFourOfAKind(hand, scoreKeeper)
            scoreKeeper[1] = 7
            return FOUR_OF_A_KIND
        elseif isFullHouse(hand, scoreKeeper)
            scoreKeeper[1] = 6
            return FULL_HOUSE
        elseif isFlush(hand, scoreKeeper)
            scoreKeeper[1] = 5
            return FLUSH
        elseif isStraight(hand, scoreKeeper)
            scoreKeeper[1] = 4
            return STRAIGHT
        elseif isThreeOfAKind(hand, scoreKeeper)
            scoreKeeper[1] = 3
            return THREE_OF_A_KIND
        elseif numOfPairs(hand, scoreKeeper) == 2
            scoreKeeper[1] = 2
            return TWO_PAIR
        elseif numOfPairs(hand, scoreKeeper) == 1
            scoreKeeper[1] = 1
            return PAIR
        elseif scoreKeeper[1] == 0
            getHighCard(hand, scoreKeeper)
            getHighCardTieBreakers(hand, scoreKeeper)
        end
        return HIGH_CARD
    end

    # Hand Rank Identifiers
    function isStraight(hand, scoreKeeper)
        ranks = [c.getRankInt() for c in hand]
        highestCard = maximum(hand)

        if isStraightRoyal(hand, scoreKeeper)
            return true
        else
            for i in 1:4
                if ranks[i] + 1 != ranks[i + 1]
                    return false
                end
            end
        end

        getHighCard(hand, scoreKeeper)
        return true
    end

    function isFlush(hand, scoreKeeper)
        for i in 1:4
            if hand[i].getSuitInt() != hand[i + 1].getSuitInt()
                return false
            end
        end
        getHighCard(hand, scoreKeeper)
        getHighCardTieBreakers(hand, scoreKeeper)
        return true
    end

    function isStraightFlush(hand, scoreKeeper)
        return isFlush(hand, scoreKeeper) && isStraight(hand, scoreKeeper)
    end

    function isStraightRoyalFlush(hand, scoreKeeper)
        loopCounter = 0
        countsPerRank = zeros(Int, 13)
        suitPerCard = zeros(Int, 5)
        cardsToCheck = [0, 9, 10, 11, 12]

        for i in 1:5
            card = hand[i]
            countsPerRank[card.getRankInt()] += 1
            suitPerCard[loopCounter] = card.getSuitInt()
            loopCounter += 1
        end

        for i in cardsToCheck
            if countsPerRank[i] != 1
                return false
            end
        end

        for i in 1:4
            if suitPerCard[i] != suitPerCard[i + 1]
                return false
            end
        end

        scoreKeeper[7] = hand[1].getSuitInt()
        return true
    end

    function isStraightRoyal(hand, scoreKeeper)
        countsPerRank = zeros(Int, 13)
        cardsToCheck = [0, 9, 10, 11, 12]
        aceSuit = 0

        for i in 1:5
            card = hand[i]
            countsPerRank[card.getRankInt()] += 1
            if card.getRankInt() == 0
                aceSuit = card.getSuitInt()
            end
        end

        for i in cardsToCheck
            if countsPerRank[i] != 1
                return false
            end
        end

        scoreKeeper[2] = 14
        scoreKeeper[7] = aceSuit
        return true
    end

    function numOfPairs(hand, scoreKeeper)
        numofPairs = 0
        iterator = 0
        kickerRank = 0
        rankofPair = zeros(Int, 2)
        countsPerRank = zeros(Int, 13)

        for i in 1:5
            countsPerRank[hand[i].getRankInt()] += 1
        end

        for i in 1:13
            if countsPerRank[i] == 2
                numofPairs += 1
                if i == 0
                    rankofPair[iterator] = 14
                else
                    rankofPair[iterator] = i
                end
                iterator += 1
            elseif countsPerRank[i] == 1
                kickerRank = i
            end
        end

        if numofPairs == 1
            ranks = [c.getRankInt() for c in hand]
            sort!(ranks)

            sKIndex = 3
            for i in 5:-1:1
                if ranks[i] != rankofPair[1]
                    scoreKeeper[sKIndex] = ranks[i]
                    sKIndex += 1
                end
            end

            scoreKeeper[2] = rankofPair[1]
        elseif numofPairs == 2
            scoreKeeper[2] = rankofPair[2]
            scoreKeeper[3] = rankofPair[1]
        end

        if numofPairs > 0
            for i in 1:5
                card = hand[i]
                if card.getRankInt() == kickerRank
                    scoreKeeper[7] = card.getSuitInt()
                end
            end
        end

        return numofPairs
    end

    function isThreeOfAKind(hand, scoreKeeper)
        countsPerRank = zeros(Int, 13)
        for i in 1:5
            card = hand[i]
            countsPerRank[card.getRankInt()] += 1
        end

        for i in 1:13
            if countsPerRank[i] == 3
                getHighCardTieBreakers(hand, scoreKeeper)
                getHighCard(hand, scoreKeeper)
                return true
            end
        end
        return false
    end

    function isFullHouse(hand, scoreKeeper)
        if isThreeOfAKind(hand, scoreKeeper) && (numOfPairs(hand, scoreKeeper) == 1)
            getHighCard(hand, scoreKeeper)
            return true
        end
        return false
    end

    function isFourOfAKind(hand, scoreKeeper)
        countsPerRank = zeros(Int, 13)
        for i in 1:5
            card = hand[i]
            countsPerRank[card.getRankInt()] += 1
        end

        for i in 1:13
            if countsPerRank[i] == 4
                getHighCard(hand, scoreKeeper)
                return true
            end
        end
        return false
    end

    function getHighCard(playerHand, scoreKeeper)
        highCard = nothing
        isAce = false
        for i in 1:5
            cardToBeChecked = playerHand[i]
    
            # Check for aces, if found only looks for other possible aces to compare
            if cardToBeChecked.getRankInt() == 0
                if isnothing(highCard)
                    highCard = cardToBeChecked
                elseif cardToBeChecked.getSuitInt() > highCard.getSuitInt()
                    highCard = cardToBeChecked
                end
                isAce = true
            # Else, find the highest rank card
            elseif isnothing(highCard) || highCard.getRankInt() < cardToBeChecked.getRankInt()
                highCard = cardToBeChecked
            # If tied for rank, compare suit
            elseif highCard.getRankInt() == cardToBeChecked.getRankInt() && highCard.getSuitInt() < cardToBeChecked.getSuitInt()
                highCard = cardToBeChecked
            end
        end
    
        # Keeps Aces High for high card
        if isAce
            scoreKeeper[1] = 14
        else
            scoreKeeper[1] = highCard.getRankInt()
        end
    
        scoreKeeper[6] = highCard.getSuitInt()
    
        return highCard
    end
    
    function getHighCardTieBreakers(hand, scoreKeeper)
        ranks = [c.getRankInt() == 0 ? 14 : c.getRankInt() for c in hand]
        sort!(ranks)  # Sorted from low to high
        sKIndex = 2
    
        # Assigns the scoreKeeper[][][3][2][1][0][]
        # Where 3-0 is:
        # ranks[3] ~ 2nd highest card, to , ranks[0] ~ lowest card
        for i in 3:-1:0
            scoreKeeper[sKIndex] = ranks[i]
            sKIndex += 1
        end
    end
    
end
end